v-next/hardhat/src/internal/cli/init/init.ts

```diff
@@ -266,21 +278,57 @@ export async function getTemplate(template?: string): Promise<Template> {
  */
 export async function validatePackageJson(
   workspace: string,
+  templatePkg: PackageJson,
   migrateToEsm?: boolean,
 ): Promise<void> {
   const absolutePathToPackageJson = path.join(workspace, "package.json");
+  const shouldUseEsm = templatePkg.type === "module";
 
   // Create the package.json file if it does not exist
   if (!(await exists(absolutePathToPackageJson))) {
     const packageJson: PackageJson = {
       name: path.basename(workspace),
-      type: "module",
       version: "1.0.0",
     };
 
+    if (shouldUseEsm) {
+      packageJson.type = "module";
+    }
+
     await writeJsonFile(absolutePathToPackageJson, packageJson);
   }
 
+  const packageManager = await getPackageManager(workspace);
+
+  // We know this works with npm, pnpm, but not with yarn. If, so we use
+  // pnpm or npm exclusively.
+  // If you read this comment and wonder if this is outdated, you can
+  // answer it by checking if the most popular versions of yarn and other
+  // package managers support `<package manager> pkg set type=module`.
+  const packageManagerToUse = packageManager === "pnpm" ? "pnpm" : "npm";
+
+  // We need to set the hardhat version in the package.json file
+  // to ensure that the template is compatible with the current Hardhat version.
+  // This is needed because we have hardhat-2 and hardhat-3 templates,
+  // and the user may install hardhat 3 first and then initialize a project
+  // with a hardhat-2 template.
+  const templateHardhatVersion = templatePkg.devDependencies?.hardhat ?? "";
```

Comment:
I believe we should be explicit about this and ask the user, like we do with the type field. Otherwise if they cancel the flow they end up with a hardhat 2 version and they may expect hardhat 3 to be installed.